File: backend/app/permissions/permission_matrix.py
"""
Permission Matrix — 5角色权限判定单一真源（后端安全边界）

Feature: platform-global-hardening

本模块是后端权限判定的**权威来源**。前端 usePermissionMatrix composable
镜像同一份定义用于 UX 优化，但不构成安全边界。

5 审计角色（映射到系统角色）：
- 审计助理 (assistant)  → system_role: auditor
- 现场经理 (manager)    → system_role: manager
- 业务合伙人 (partner)  → system_role: partner
- 质量控制复核合伙人 (qc_partner) → system_role: qc
- EQCR 技术复核人 (eqcr) → system_role: eqcr

前后端共享同一组 operation codes（与 permission_matrix_service.py 一致）。
本模块在其基础上提供面向审计 5 角色的 `can()` 函数，作为统一入口。
"""

from __future__ import annotations

from typing import Any

from app.services.permission_matrix_service import (
    OPERATION_CODES,
    can as _matrix_can,
    get_allowed_operations,
)

# ─── 5 审计角色常量 ──────────────────────────────────────────────────────────
AUDIT_ROLES = ("assistant", "manager", "partner", "qc_partner", "eqcr")

# ─── 角色别名→系统角色映射（归一化） ──────────────────────────────────────────
ROLE_ALIAS_MAP: dict[str, str] = {
    # 审计 5 角色 → 系统角色
    "assistant": "auditor",
    "auditor": "auditor",
    "manager": "manager",
    "partner": "partner",
    "signing_partner": "partner",
    "qc_partner": "qc",
    "qc": "qc",
    "qc_reviewer": "qc",
    "quality_control": "qc",
    "eqcr": "eqcr",
    # admin 保留
    "admin": "admin",
}

# ─── 资源→操作前缀映射 ────────────────────────────────────────────────────────
RESOURCE_ACTION_MAP: dict[str, dict[str, str]] = {
    "workpaper": {
        "edit": "wp:edit",
        "review": "wp:review",
        "read": "project:view",
    },
    "report": {
        "edit": "report:edit",
        "sign": "report:sign",
        "read": "project:view",
    },
    "note": {
        "edit": "note:edit",
        "read": "project:view",
    },
    "project": {
        "view": "project:view",
        "read": "project:view",
    },
    "archive": {
        "manage": "archive:manage",
    },
}

# ─── 锁定底稿限制 ────────────────────────────────────────────────────────────
# 底稿锁定后，只有 partner/admin 可以解锁/编辑
LOCKED_OVERRIDE_ROLES = {"admin", "partner"}

# ─── 复核后限制 ───────────────────────────────────────────────────────────────
# 已复核底稿，编辑操作降级为只读（除非是 partner/admin 重新打开）
POST_REVIEW_EDIT_ROLES = {"admin", "partner", "manager"}


def normalize_role(role: str) -> str:
    """将审计角色名/别名归一化为系统角色名。

    Args:
        role: 原始角色字符串（如 "assistant", "qc_partner", "signing_partner"）

    Returns:
        系统角色名（如 "auditor", "qc", "partner"）
    """
    r = role.lower().strip()
    return ROLE_ALIAS_MAP.get(r, r)


def can(
    role: str,
    action: str,
    resource: str,
    context: dict[str, Any] | None = None,
) -> bool:
    """统一权限判定入口 — 镜像前端 usePermissionMatrix.can()。

    后端安全边界：此函数的判定结果为最终权威。前端同名函数仅用于 UX。

    Args:
        role: 审计角色名（5角色别名或系统角色名均可）
        action: 操作动作（如 "edit", "review", "sign", "read"）
        resource: 资源类型（如 "workpaper", "report", "note", "project"）
        context: 可选上下文（支持 locked/reviewed/project_role 等）

    Returns:
        True 表示允许，False 表示拒绝

    Examples:
        >>> can("assistant", "edit", "workpaper")
        True
        >>> can("eqcr", "edit", "workpaper")
        False
        >>> can("assistant", "edit", "workpaper", {"locked": True})
        False
    """
    ctx = context or {}
    system_role = normalize_role(role)
    project_role = ctx.get("project_role")

    # ─── 直接用 operation code（兼容已有端点直接传 operation） ─────────
    if ":" in action and resource == "":
        operation = action
    else:
        # ─── 从 resource + action 解析 operation code ─────────────────
        resource_map = RESOURCE_ACTION_MAP.get(resource, {})
        operation = resource_map.get(action)
        if operation is None:
            # 未映射的 action:resource 组合 → 拒绝（安全侧）
            return False

    # ─── 基础权限判定（委托 permission_matrix_service） ────────────────
    if not _matrix_can(system_role, project_role, operation):
        return False

    # ─── 上下文约束：锁定底稿 ─────────────────────────────────────────
    if ctx.get("locked") and action == "edit":
        if system_role not in LOCKED_OVERRIDE_ROLES:
            return False

    # ─── 上下文约束：已复核底稿 ───────────────────────────────────────
    if ctx.get("reviewed") and action == "edit":
        if system_role not in POST_REVIEW_EDIT_ROLES:
            return False

    return True


def get_operations_for_role(
    role: str,
    project_role: str | None = None,
) -> set[str]:
    """返回指定角色的允许操作集合。

    Args:
        role: 审计角色名
        project_role: 可选项目职责

    Returns:
        允许的 operation code 集合
    """
    system_role = normalize_role(role)
    return get_allowed_operations(system_role, project_role)


def get_all_operation_codes() -> list[str]:
    """返回全部已定义的 operation codes。"""
    return list(OPERATION_CODES)

File: audit-platform/frontend/src/permissions/permission-matrix.ts
/**
 * Permission Matrix — 单一权限定义（前后端共享）
 * Feature: platform-global-hardening
 *
 * 5 角色项目级权限区分，单一入口 `can(role, action, resource, context)`。
 * 前端仅用于体验优化（按钮显隐/只读态），后端 deps 层为安全边界。
 *
 * 角色映射（中文→code）：
 * - 审计助理 = assistant
 * - 现场经理 = manager
 * - 业务合伙人 = partner
 * - 质量控制复核合伙人 = qc_partner
 * - EQCR 技术复核人 = eqcr
 */

// ─── 类型定义 ─────────────────────────────────────────────────────────────────

export type Role = 'assistant' | 'manager' | 'partner' | 'qc_partner' | 'eqcr'

export type Action = 'edit' | 'review' | 'approve' | 'delete' | 'export' | 'import' | 'refresh' | 'view'

export type Resource = 'workpaper' | 'adjustment' | 'report' | 'note' | 'template' | 'project_settings'

export interface PermissionContext {
  projectId?: string
  wpCode?: string
  /** 底稿是否已锁定（仅 partner 可解锁） */
  isLocked?: boolean
  /** 复核状态：approved = 审批后不可编辑 */
  reviewStatus?: string
}

// ─── 角色→资源→动作 权限定义表 ────────────────────────────────────────────────

/**
 * 基础权限矩阵：role → resource → allowed actions
 *
 * 规则来源（design.md + requirements 7.6）：
 * - assistant: edit workpaper, view all, export, import
 * - manager: all assistant + review, approve adjustments, edit project_settings
 * - partner: all manager + approve reports, refresh, delete
 * - qc_partner: view all, review all, cannot edit
 * - eqcr: view all, review (read-only), cannot edit/approve
 */
const PERMISSION_TABLE: Record<Role, Record<Resource, Set<Action>>> = {
  assistant: {
    workpaper: new Set(['edit', 'view', 'export', 'import']),
    adjustment: new Set(['edit', 'view', 'export', 'import']),
    report: new Set(['view', 'export']),
    note: new Set(['edit', 'view', 'export']),
    template: new Set(['view']),
    project_settings: new Set(['view']),
  },
  manager: {
    workpaper: new Set(['edit', 'review', 'view', 'export', 'import']),
    adjustment: new Set(['edit', 'review', 'approve', 'view', 'export', 'import']),
    report: new Set(['edit', 'review', 'view', 'export']),
    note: new Set(['edit', 'review', 'view', 'export']),
    template: new Set(['edit', 'view']),
    project_settings: new Set(['edit', 'view']),
  },
  partner: {
    workpaper: new Set(['edit', 'review', 'approve', 'delete', 'view', 'export', 'import', 'refresh']),
    adjustment: new Set(['edit', 'review', 'approve', 'delete', 'view', 'export', 'import']),
    report: new Set(['edit', 'review', 'approve', 'delete', 'view', 'export', 'refresh']),
    note: new Set(['edit', 'review', 'approve', 'delete', 'view', 'export', 'refresh']),
    template: new Set(['edit', 'delete', 'view']),
    project_settings: new Set(['edit', 'delete', 'view']),
  },
  qc_partner: {
    workpaper: new Set(['review', 'view', 'export']),
    adjustment: new Set(['review', 'view', 'export']),
    report: new Set(['review', 'view', 'export']),
    note: new Set(['review', 'view', 'export']),
    template: new Set(['view']),
    project_settings: new Set(['view']),
  },
  eqcr: {
    workpaper: new Set(['review', 'view', 'export']),
    adjustment: new Set(['review', 'view', 'export']),
    report: new Set(['review', 'view', 'export']),
    note: new Set(['review', 'view', 'export']),
    template: new Set(['view']),
    project_settings: new Set(['view']),
  },
}

// ─── 全部角色列表（用于属性测试枚举） ──────────────────────────────────────────
export const ALL_ROLES: readonly Role[] = ['assistant', 'manager', 'partner', 'qc_partner', 'eqcr'] as const
export const ALL_ACTIONS: readonly Action[] = ['edit', 'review', 'approve', 'delete', 'export', 'import', 'refresh', 'view'] as const
export const ALL_RESOURCES: readonly Resource[] = ['workpaper', 'adjustment', 'report', 'note', 'template', 'project_settings'] as const

// ─── 核心判定函数 ─────────────────────────────────────────────────────────────

/**
 * 单一入口：判断指定角色是否可对指定资源执行指定动作。
 *
 * 前后端共享此定义——前端用于体验优化（按钮显隐/只读），后端 deps 层为安全边界。
 *
 * 特殊规则：
 * 1. 锁定底稿：仅 partner 可解锁（执行 edit 动作）
 * 2. 复核通过后（reviewStatus === 'approved'）：任何角色均不可 edit
 * 3. view 动作对所有角色始终允许（信息透明）
 *
 * @param role - 当前用户角色
 * @param action - 请求的动作
 * @param resource - 目标资源类型
 * @param context - 可选上下文（锁定状态/复核状态等）
 * @returns 是否允许执行
 */
export function can(
  role: Role,
  action: Action,
  resource: Resource,
  context?: PermissionContext,
): boolean {
  // view 始终允许（信息透明原则）
  if (action === 'view') return true

  // 特殊规则 2：复核通过后不可编辑（任何角色）
  if (context?.reviewStatus === 'approved' && action === 'edit') {
    return false
  }

  // 特殊规则 1：锁定底稿仅 partner 可操作 edit
  if (context?.isLocked && resource === 'workpaper' && action === 'edit') {
    return role === 'partner'
  }

  // 基础权限表查询
  const rolePermissions = PERMISSION_TABLE[role]
  if (!rolePermissions) return false

  const resourcePermissions = rolePermissions[resource]
  if (!resourcePermissions) return false

  return resourcePermissions.has(action)
}

/**
 * 判断角色对指定资源是否只读（不可 edit）。
 * 等价于 `!can(role, 'edit', resource, context)`
 */
export function isReadonly(
  role: Role,
  resource: Resource,
  context?: PermissionContext,
): boolean {
  return !can(role, 'edit', resource, context)
}

/**
 * 获取角色对指定资源的所有允许动作集合。
 */
export function getAllowedActions(
  role: Role,
  resource: Resource,
  context?: PermissionContext,
): Action[] {
  return ALL_ACTIONS.filter(action => can(role, action, resource, context))
}

/**
 * 返回拒绝原因，允许则返回 null。
 */
export function whyCannotDo(
  role: Role,
  action: Action,
  resource: Resource,
  context?: PermissionContext,
): string | null {
  if (can(role, action, resource, context)) return null

  if (context?.reviewStatus === 'approved' && action === 'edit') {
    return `复核已通过，${resource} 不可编辑`
  }

  if (context?.isLocked && resource === 'workpaper' && action === 'edit') {
    return `底稿已锁定，仅业务合伙人可解锁编辑`
  }

  const roleNames: Record<Role, string> = {
    assistant: '审计助理',
    manager: '现场经理',
    partner: '业务合伙人',
    qc_partner: '质量控制复核合伙人',
    eqcr: 'EQCR技术复核人',
  }

  return `${roleNames[role]}无权对${resource}执行${action}操作`
}
